File: Esercitazione4/esercizio5/QuickSort.java
package esercizio5;

import java.util.Random;

public class QuickSort implements SortingAlgorithm{
	Random generatore = new Random();
	@Override
	public void sort(int[] a) {
		int n = a.length;
		qsort(a,0,n-1);	
	}
	private void qsort(int[]a, int inf, int sup){
			int pivot = a[inf];
			scambia(a,sup,pivot);
			int p=a[sup];
			int i = inf;
			int j=sup-1;
			while(i<=j){	
				while(i<=j && a[j]>=p) i++;
				while(i<=j && a[j] >=p) j--;
				if(i<j) scambia (a,i,j);
			}
			scambia(a,i,sup);		
			if (inf <sup) {;
			qsort(a,inf,i-1);	
			qsort(a,i+1,sup);
			}
	}
	private void scambia(int[]a,int i,int j){
		int x=a[i];
		a[i]=a[j];
		a[j]=x;
	}
	
	
}

File: Esercitazione4/sabatino/Esercizio 6/QuickSort.c
/*
 * quickSort.c
 *
 *  Created on: 14/dic/2013
 *      Author: Mauro
 */
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <sys/time.h>

#include "SortingAlgorithm.h"

	void QSort(int a[],int n){
		QuickSortMidToExamine(a,0,n-1);
	}

	void QuickSortMidToExamine(int a[],int inf, int sup){
		if(inf>=sup) return;
		int p = QuickSortMidToExaminePartition(a,inf,sup);
		QuickSortMidToExamine(a,inf,p-1);
		QuickSortMidToExamine(a,p+1,sup);
	}
	int QuickSortMidToExaminePartition(int a[],int inf, int sup){
		int pivot = a[inf];
		int i = inf+1;
		int j = sup;
		while(i<=j){
			if(a[i]<pivot)	i++;
			else scambia(a,i,j--);
		}
		scambia(a,inf,j);
		return j ;
	}

	 void scambia(int a[],int i,int j){
			int x=a[i];
			a[i]=a[j];
			a[j]=x;
		}
